### PR TITLE
Refactory/using complex interaction template

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
Revamping the fluid relevant class and method.
Basically, I have used partial template specialization to identify inner, contact and contact wall interactions.
I have revised the template complex interaction class to combine two or more interaction together.
However, the changes related to solid dynamics are minimum.

### CODE REVIEW

**Comment by ChiZhangatTUM** [comment]
> Revamping the fluid relevant class and method. Basically, I have used partial template specialization to identify inner, contact and contact wall interactions. I have revised the template complex interaction class to combine two or more interaction together. However, the changes related to solid dynamics are minimum.

Great work. Would do you please add more principles underlying this refactoring when you have time? Would be very help to review this PR.

**Comment by ChiZhangatTUM** [comment]
Subdirectory "modules" is added? Which kind of modules should be belonged to this sub?

**Comment by ChiZhangatTUM** [comment]
How about a meeting to briefly summarize the underlying consideration with this PR and all reviewers are kindly invited?

File: src/shared/bodies/complex_bodies/unstructured_mesh.h [hunk 1]
```diff
@@ -1,71 +1,65 @@
...
 using namespace std;
 namespace SPH
 {
 /**
- * @class readMeshFile
+ * @class ANSYSMesh
  * @brief ANASYS mesh.file parser class
  */
-class readMeshFile
+class ANSYSMesh
```
Comment (ChiZhangatTUM) [inline src/shared/bodies/complex_bodies/unstructured_mesh.h:43]: ANSYSMesh @ZhentongWang why the mesh name is with "ANSYS"?
'

**Comment by Xiangyu-Hu** [comment]
> How about a meeting to briefly summarize the underlying consideration with this PR and all reviewers are kindly invited?

Sure. However, this pull request is only conservative modification, that is, it does not change the any currently test functions. There is nothing new in the pull request.

File: src/shared/bodies/complex_bodies/unstructured_mesh.h (hunk 1 above)
Comment (Xiangyu-Hu) [reply src/shared/bodies/complex_bodies/unstructured_mesh.h:43]: It is the file format used by ANSYS, I am not it is also used by other code.

**Comment by ChiZhangatTUM** [comment]
> > How about a meeting to briefly summarize the underlying consideration with this PR and all reviewers are kindly invited?
> 
> Sure. However, this pull request is only conservative modification, that is, it does not change the any currently test functions. There is nothing new in the pull request.

Understood. A meeting would be very helpful for other developers to follow your idea and to enhance the community building of SPHinXsys. If possible, we may consider have a routine meeting, for example monthly or bi-monthly,  for the main contributors who are willing to review PR, solve Issue and refactor the code.

**Comment by ChiZhangatTUM** [comment]
@Shuaihao-Zhang @Bo-Zhang1995 @junwei-jiang @FabienPean-Virtonomy @DongWuTUM I have asked Dr. Hu to give us a brief introduction on the underlying consider of this PR in the afternoon 4pm this Wednesday, please join us if you have time. I will send you the link later.

**Comment by ChiZhangatTUM** [comment]
@Shuaihao-Zhang @Bo-Zhang1995 @junwei-jiang @FabienPean-Virtonomy @yuanjiajy @DongWuTUM 
Please join the meeting with the following:
Join Zoom Meeting
https://tum-conf.zoom-x.de/j/2431421170

Meeting ID: 243 142 1170

File: README.md
Comment (ChiZhangatTUM) [inline README.md:1]: Maybe we can change the section title of "Journal publications" to "References".

File: src/for_2D_build/bodies/unstructured_mesh.cpp [hunk 2]
```diff
@@ -1,19 +1,68 @@
...
-#include "common_shared_FVM_classes.h"
+#include "unstructured_mesh.h"
+
 namespace SPH
 {
 //=================================================================================================//
-void readMeshFile::getDataFromMeshFile()
+ANSYSMesh::ANSYSMesh(const std::string &full_path)
+{
+    getDataFromMeshFile(full_path);
```
Comment (ChiZhangatTUM) [inline src/for_2D_build/bodies/unstructured_mesh.cpp:9]: this function name may change to "readANSYSMesh" or "loadANSYSMesh" .

File: src/for_2D_build/bodies/unstructured_mesh.cpp [hunk 3]
```diff
@@ -1,19 +1,68 @@
...
+    std::string y_coordinate_string = y_part.erase(0, divide_position);
+
+    Vec2d coordinate = Vec2d::Zero();
+    std::istringstream stream_x, stream_y;
+    stream_x.str(x_coordinate_string);
+    stream_y.str(y_coordinate_string);
+    stream_x >> coordinate[0];
+    stream_y >> coordinate[1];
+    node_coordinates.push_back(coordinate);
+}
```
Comment (ChiZhangatTUM) [inline src/for_2D_build/bodies/unstructured_mesh.cpp:29]: how about more compact way of 

size_t divide_position = text_line.find_first_of(" ");
Vec2d coordinate = Vec2d::Zero();
std::istringstream( text_line.erase(divide_position) ) >> coordinate[0];
std::istringstream( text_line.erase(0, divide_position) ) >> coordinate[1];

Comment (Xiangyu-Hu) [reply src/for_2D_build/bodies/unstructured_mesh.cpp:29]: @ZhentongWang Could you consider Chi comment for a later revision of the code?

File: src/shared/particle_dynamics/base_local_dynamics.h [hunk 4]
```diff
@@ -228,51 +302,56 @@ class Average : public ReduceSumType
...
+struct ConstructorArgs
 {
     BodyRelationType &body_relation_;
     std::tuple<OtherArgs...> others_;
-    LocalDynamicsParameters(BodyRelationType &body_relation, OtherArgs &&...other_args)
-        : body_relation_(body_relation), others_(std::forward<OtherArgs>(other_args)...){};
+    SPHBody &getSPHBody() { return body_relation_.getSPHBody(); };
+    ConstructorArgs(BodyRelationType &body_relation, OtherArgs... other_args)
+        : body_relation_(body_relation),
+          others_(other_args...){};
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply src/shared/particle_dynamics/base_local_dynamics.h:316]: before, this forwarding seems not working, I can have a try again.

[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply src/shared/particle_dynamics/base_local_dynamics.h:316]: It still not working and show the following error:
/home/xiangyu/SPHinXsys/src/shared/particle_dynamics/general_dynamics/kernel_correction.h:64:81: note: passing argument to parameter 'parameters' here
    explicit KernelCorrectionMatrix(ConstructorArgs<BodyRelationType, FirstArg> parameters)

File: README.md
Comment (Xiangyu-Hu) [reply README.md:1]: This will be update from documentation next time.

**Review by Lisu-lisa — APPROVED** [review]

File: src/shared/particle_dynamics/base_local_dynamics.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (FabienPean-Virtonomy) [reply src/shared/particle_dynamics/base_local_dynamics.h:316]: This is not an error, nor a full error log, so I can't say what the error could be about.

[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply src/shared/particle_dynamics/base_local_dynamics.h:316]: Sorry. I have only show a piece of the message. The code is not able to compile and I am not able to find a clue to correct it.

**Pull request merged**